### ISSUE TITLE
`test_iface_binary_bcast` failing on AArch64

### ISSUE THREAD

**Issue by nSircombe** [oneapi-src/oneDNN#1125]
# Summary
The [recently added ](https://github.com/oneapi-src/oneDNN/commit/e3a9759530cec4e195ae4baabd5cf192e0340ecf)`test_iface_binary_bcast` test fails on AArch64 builds.

# Version
Current head of master (https://github.com/oneapi-src/oneDNN/commit/bdc1d6048e6ba4d1c931c66d6592b13fd09b4883). Failures start with the integration of [`test_iface_binary_bcast`](https://github.com/oneapi-src/oneDNN/commit/e3a9759530cec4e195ae4baabd5cf192e0340ecf).

# Environment
 - Arm Neoverse-N1 (AArch64, Arm-v8.2a)
 - GCC 10.2.0
 - RHEL 8

# Steps to reproduce

```
mkdir -p build
cd build
cmake -DDNNL_CPU_RUNTIME=OMP -DDNNL_BUILD_FOR_CI=ON ../. && make -j 32
ctest -R test_iface_binary_bcast  --verbose
```

# Observed behavior
All the the 'CPU not optimised' tests (https://github.com/oneapi-src/oneDNN/blob/bdc1d6048e6ba4d1c931c66d6592b13fd09b4883/tests/gtests/test_iface_binary_bcast.cpp#L167) are failing.

```
test 55                                                                                                                                                                                                                              [45/1990]
    Start 55: test_iface_binary_bcast                                                                                                                                                                                                         
                                                                                                                                                                                                                                              
55: Test command: /home/natsir01/oneDNN-gerrit/build/tests/gtests/test_iface_binary_bcast                                                                                                                                                     
55: Test timeout computed to be: 10000000                                                                                                                                                                                                     
55: Note: Google Test filter = *:-*_GPU*                                                                                                                                                                                                      
55: [==========] Running 30 tests from 2 test suites.                                                                                                                                                                                         
55: [----------] Global test environment set-up.                                                                                                                                                                                              
55: [----------] 25 tests from CPUOptimizedDims/binary_bcast_test_t                                                                                                                                                                           
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/0                                                                                                                                                          
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/0 (2 ms)                                                                                                                                                   
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/1                                                                                                                                                          
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/1 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/2
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/2 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/3
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/3 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/5
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/5 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/6
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/6 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/7
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/7 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/8
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/8 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/9
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/9 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/10
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/10 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/11
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/11 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/12
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/12 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/13
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/13 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/14
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/14 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/15
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/15 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/16
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/16 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/17
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/17 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/18
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/18 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/19
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/19 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/20
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/20 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/21
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/21 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/22
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/22 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/23
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/23 (0 ms)
55: [ RUN      ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/24
55: [       OK ] CPUOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/24 (0 ms)
55: [----------] 25 tests from CPUOptimizedDims/binary_bcast_test_t (2 ms total)
55:
55: [----------] 5 tests from CPUNotOptimizedDims/binary_bcast_test_t
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/0
55: /home/natsir01/oneDNN-gerrit/tests/gtests/test_iface_binary_bcast.cpp:94: Failure
55: Expected: (impl_info_no_bcast) != (impl_info_bcast), actual: "ref:any" vs "ref:any"
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/0, where GetParam() = (4-byte object <01-00 00-00>, { 1, 1, 8, 1, 8 }, false) (0 ms)                                                                   
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/1
55: /home/natsir01/oneDNN-gerrit/tests/gtests/test_iface_binary_bcast.cpp:94: Failure
55: Expected: (impl_info_no_bcast) != (impl_info_bcast), actual: "ref:any" vs "ref:any"
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/1, where GetParam() = (4-byte object <01-00 00-00>, { 1, 8, 1, 1, 8 }, false) (0 ms)                                                                   
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/2
55: /home/natsir01/oneDNN-gerrit/tests/gtests/test_iface_binary_bcast.cpp:94: Failure
55: Expected: (impl_info_no_bcast) != (impl_info_bcast), actual: "ref:any" vs "ref:any"
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/2, where GetParam() = (4-byte object <01-00 00-00>, { 8, 1, 1, 1, 1 }, false) (0 ms)                                                                   
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/3
55: /home/natsir01/oneDNN-gerrit/tests/gtests/test_iface_binary_bcast.cpp:94: Failure
55: Expected: (impl_info_no_bcast) != (impl_info_bcast), actual: "ref:any" vs "ref:any"
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/3, where GetParam() = (4-byte object <01-00 00-00>, { 1, 1, 8, 1 }, false) (0 ms)                                                                      
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4
55: /home/natsir01/oneDNN-gerrit/tests/gtests/test_iface_binary_bcast.cpp:94: Failure
55: Expected: (impl_info_no_bcast) != (impl_info_bcast), actual: "ref:any" vs "ref:any"
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4, where GetParam() = (4-byte object <01-00 00-00>, { 8, 1, 1 }, false) (0 ms)                                                                         
55: [----------] 5 tests from CPUNotOptimizedDims/binary_bcast_test_t (0 ms total)
55:
55: [----------] Global test environment tear-down
55: [==========] 30 tests from 2 test suites ran. (2 ms total)
55: [  PASSED  ] 25 tests.
55: [  FAILED  ] 5 tests, listed below:
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/0, where GetParam() = (4-byte object <01-00 00-00>, { 1, 1, 8, 1, 8 }, false)                                                                          
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/1, where GetParam() = (4-byte object <01-00 00-00>, { 1, 8, 1, 1, 8 }, false)                                                                          
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/2, where GetParam() = (4-byte object <01-00 00-00>, { 8, 1, 1, 1, 1 }, false)                                                                          
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/3, where GetParam() = (4-byte object <01-00 00-00>, { 1, 1, 8, 1 }, false)                                                                             
55: [  FAILED  ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4, where GetParam() = (4-byte object <01-00 00-00>, { 8, 1, 1 }, false)                                                                                
55:
55:  5 FAILED TESTS
1/1 Test #55: test_iface_binary_bcast ..........***Failed    0.01 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.02 sec

The following tests FAILED:
         55 - test_iface_binary_bcast (Failed)
Errors while running CTest
Output from these tests are in: /home/natsir01/oneDNN-gerrit/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

# Expected behavior
All tests passing.

**Comment by dzarukin** [comment]
Hi @nSircombe, thank you for reporting the issue.

Assigning it to the author of the test. @bartekxk, could you, please, take a look on this one? Thank you.

**Comment by nSircombe** [comment]
Cheers @dzarukin.
@bartekxk, I forgot to add above that `impl_info_no_bcast` and `impl_info_bcast` return `ref:any` in all cases on our AArch64 builds.

**Comment by igorsafo** [comment]
Hi @nSircombe ,
the fix is available in master: e05681e
Could you please validate it on your side?

**Comment by nSircombe** [comment]
Hi @igorsafo,
That's sorted it...

```
....
55: [ RUN      ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4
55: [  SKIPPED ] Binary impl_info_str should be same only on x64 CPU
55: [       OK ] CPUNotOptimizedDims/binary_bcast_test_t.TestBinaryOptimizedBroadcast/4 (0 ms)
55: [----------] 5 tests from CPUNotOptimizedDims/binary_bcast_test_t (0 ms total)
55:
55: [----------] Global test environment tear-down
55: [==========] 30 tests from 2 test suites ran. (1 ms total)
55: [  PASSED  ] 30 tests.
1/1 Test #55: test_iface_binary_bcast ..........   Passed    0.01 sec

The following tests passed:
        test_iface_binary_bcast

100% tests passed, 0 tests failed out of 1
```

Thanks for your help.
